drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,958 @@
+	struct spi_engine_cmd_queue *local_fifo;
+
+	local_fifo = *fifo;
+	while (local_fifo->next != NULL) {
+		// Get the last element
+		local_fifo = local_fifo->next;
+	}
+
+	// Create a new element
+	spi_engine_queue_new_cmd(&to_add, cmd);
+	// Add as the next element
+	local_fifo->next = to_add;
+}
+
+/**
+ * @brief Add a command at the beginning of an existing queue
+ *
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Command to be added
+ */
+void spi_engine_queue_append_cmd(struct spi_engine_cmd_queue **fifo,
+				 uint32_t cmd)
+{
+	struct spi_engine_cmd_queue *to_add = NULL;
+
+	// Create a new element
+	spi_engine_queue_new_cmd(&to_add, cmd);
+	// Interchange the addresses
+	to_add->next = *fifo;
+	*fifo = to_add;
+}
+
+/**
+ * @brief Get the command from the end of the queue
+ *
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Value of the command that was extracted
+ * @return int32_t This function allways return SUCCESS
+ */
+int32_t spi_engine_queue_fifo_get_cmd(struct spi_engine_cmd_queue **fifo,
+				      uint32_t *cmd)
+{
+	struct spi_engine_cmd_queue *local_fifo;
+	struct spi_engine_cmd_queue *last_fifo = NULL;
+
+	local_fifo = *fifo;
+	while (local_fifo->next != NULL) {
+		// Get the last element
+		last_fifo = local_fifo;
+		local_fifo = local_fifo->next;
+	}
+	*cmd = local_fifo->cmd;
+	if(last_fifo->next != NULL) {
+		// Remove the last element
+		free(last_fifo->next);
+		last_fifo->next = NULL;
+	} else {
+		// Delete the queue
+		free(last_fifo);
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the command from the beginning of the queue
+ *
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Value of the command that was extracted
+ * @return int32_t Return FAILURE if the queue is empty
+ */
+int32_t spi_engine_queue_lifo_get_cmd(struct spi_engine_cmd_queue **fifo,
+				      uint32_t *cmd)
+{
+	struct spi_engine_cmd_queue *local_fifo;
+
+
+	if ((*fifo)->next) {
+		local_fifo = (*fifo);
+		*cmd = local_fifo->cmd;
+		*fifo = local_fifo->next;
+		free(local_fifo);
+	} else {
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Free the memory allocated by a queue
+ *
+ * @param fifo The queue that needs it's memory freed
+ * @return int32_t This function allways return SUCCESS
+ */
+int32_t spi_engine_queue_free(struct spi_engine_cmd_queue **fifo)
+{
+	if ((*fifo)->next)
+		spi_engine_queue_free(&(*fifo)->next);
+	if((*fifo) != NULL) {
+		free(*fifo);
+		*fifo = NULL;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write the SPI engine's command fifo
+ *
+ * @param desc Decriptor containing SPI Engine's parameters
+ * @param cmd The command that will be written
+ * @return int32_t Returns FAILURE if the the axi transfer failed
+ */
+int32_t spi_engine_write_cmd_reg(struct spi_engine_desc *desc,
+				 uint32_t cmd)
+{
+	int32_t ret;
+
+	/* Check if offload is enabled */
+	if(desc->offload_config & (OFFLOAD_TX_EN | OFFLOAD_RX_EN)) {
```

Comment:
`desc` should be tested before dereferencing. This applies to all subsequent functions that use desc as parameter.